gravitee-io/gravitee-api-management
# test(grpc): migrate standalone grpc tests to gateway-testing-sdk

## Issue

https://gravitee.atlassian.net/browse/APIM-432

## Description

### Testing SDK
- add a URLStreamHandler to accept grpc protocol (used to override endpoint port)
- add an AbstractGrpcGatewayTest configuring vertx server (non reactive) and provides utility functions to create service and channel
- clear system properties set by developer thanks to GatewayConfigurationBuilder on server stops. It avoids to keep undesirable properties for following tests
- check existence of endpoints before operating on it (AbstractGatewayTest#updateEndpoints)
- add some utility functions in AbstractGatewayTest and GatewayConfigurationBuilder

### GRPC integration test
- remove the tests from gateway-standalone-container
- migrate them using the gateway testing sdk, with:
  - v3 gateway
  - jupiter gateway, jupiter api
  - jupiter gateway, v3 api

⚠️ [...ner/src/test/java/io/gravitee/gateway/standalone/grpc/GrpcSecuredServerStreamingTest.java](https://github.com/gravitee-io/gravitee-api-management/compare/apim-432-verify-grpc?expand=1#diff-91de28f0b56cffbebe428d3a620be461bd7228f3ad9557c7235f0596ae18269d) was disabled because of race conditions.
I re-enabled it because the gateway testing sdk should handle properly the case:
- system properties are clean on test class stop (top be precise, when vertx stops)
- `@GatewayTest` is a meta-annotation using `@ResourceLock(value = Resources.SYSTEM_PROPERTIES, mode = ResourceAccessMode.READ_WRITE)`. Its purpose is to "lock" the access to a particular resource (here `SYSTEM_PROPERTIES`) preventing to test to modify it at the same time

---

📚&nbsp;&nbsp;View the storybook of this branch [here](https://612657caa8e859003a8a6430-irvhqcbbrn.chromatic.com)


🚀 CI was able to deploy the build of this PR, so you can now try it directly [here](https://apimnightlywebui24386.z6.web.core.windows.net/apim-432-verify-grpc/index.html)
_Notes_: The deployed app is linked to the management API of APIM master. (Same login and password as APIM master)
